# 03. Data Preprocessing - Automobile Loan Default Prediction
**Module:** IT3051 – Fundamentals of Data Mining
**Objective:** Turn the raw, messy dataset into a clean, encoded, split, model-ready dataset, implementing every decision documented in `02_eda.ipynb`'s Section 11 summary.

---

### Recap of the plan (from `02_eda.ipynb` Section 11)
1. Setup - reload the raw data fresh (this notebook builds its own self-contained pipeline).
2. Fix data types (numeric-looking text columns).
3. Apply rule-based placeholder/sentinel fixes (`Employed_Days`, `XNA`, `"##"`, `Score_Source_2`, `Population_Region_Relative`), drop `ID`.
4. Stratified train/test split - **done before any step that learns a statistic from the data**, to avoid leakage.
5. Handle missing values (fit on training data only).
6. Outlier treatment (log-transform `Client_Income`).
7. Encode categorical variables (fit on training data only).
8. Feature scaling (fit on training data only).
9. Save processed train/test sets to `data/processed/`.
10. Summary of preprocessing decisions.

Rare-category grouping and new engineered features are deliberately left for `04_feature_engineering.ipynb`.


## 1. Setup

We reload the raw dataset fresh, independently of `02_eda.ipynb`'s exploration copy. This notebook is meant to be a self-contained, reusable pipeline — not a continuation of the EDA notebook's in-memory state — so that it can be re-run on its own (e.g. from the command line) to reproduce `data/processed/` from scratch.


In [1]:
import os
import numpy as np
import pandas as pd

pd.set_option('display.max_columns', 50)
pd.set_option('display.width', 1000)

print("Libraries imported successfully!")

Libraries imported successfully!


In [2]:
# Relative path handling: works whether launched from the workspace root or inside the notebooks folder
data_path = 'data/raw/Train_Dataset.csv' if os.path.exists('data/raw/Train_Dataset.csv') else '../data/raw/Train_Dataset.csv'

df = pd.read_csv(data_path, low_memory=False)

print(f"Dataset loaded successfully from: '{data_path}'")
print(f"Shape: {df.shape}")

Dataset loaded successfully from: '../data/raw/Train_Dataset.csv'
Shape: (121856, 40)


**Confirms:** shape should be `(121856, 40)`, matching both `01_data_understanding.ipynb` and `02_eda.ipynb` — this is the same untouched raw file, giving this notebook a clean, independent starting point to build the actual preprocessing pipeline from.
